dashboard/app/models/concerns/partial_registration.rb

```diff
@@ -7,62 +7,53 @@
 module PartialRegistration
   extend ActiveSupport::Concern
 
-  USER_ATTRIBUTES_SESSION_KEY = 'devise.user_attributes'
-  OAUTH_PARAMS_TO_STRIP = %w(oauth_token oauth_refresh_token)
+  SESSION_KEY = 'partial_registration'
 
   module ClassMethods
-    def new_from_partial_registration(session)
+    def new_from_partial_registration(session, &block)
       raise 'No partial registration was in progress' unless PartialRegistration.in_progress? session
```

Comment:
Is this the error we would present to the user in the rare scenario where memcached has evicted their cache entry? If so, should we tell them the error was temporary and that they should start over?

Reply:
I decided that handling an evicted registration attempt is out of scope for this model initializer method; that's a controller concern.  The guard clause and error are here to prevent misuse of this method. the caller is responsible for only calling `new_from_partial_registration` when a partial registration is in progress.

As currently implemented, that is the case. The only call into this method is from `User.new_with_session`, which itself contains a guard checking that a partial registration is in progress and falling back on [the Devise `new_with_session` implementation](https://www.rubydoc.info/github/plataformatec/devise/Devise/Models/Registerable/ClassMethods), which is equivalent to `new`.

https://github.com/code-dot-org/code-dot-org/blob/4267146cc5d1c47cd729951ab77a5d9ea406421b/dashboard/app/models/user.rb#L778-L783

`new_with_session`, in turn, is only called from RegistrationsController#new (`GET /users/sign_up`), and only when a partial registration is in progress.  Otherwise, we render the initial sign up page, starting the process over from the beginning.

https://github.com/code-dot-org/code-dot-org/blob/2c57517812b55936f180de942181455deea8f603/dashboard/app/controllers/registrations_controller.rb#L16-L28

So in almost all cases where a partial registration has been evicted from the cache, the user will be seamlessly taken back to the beginning of the registration process.

In the exceedingly rare case where a a registration is evicted from the cache between execution of RegistrationsController#new and User#new_from_partial_registration (and assuming the value is not itself cached in local memory from the `PartialRegistration.in_progress?` check), this error would be uncaught and result in a generic 500 "sad bee" for the user.  The 'No partial registration was in progress' langauge would show up in Honeybadger, for our use.